nixos/modules/services/hardware/monado.nix

```diff
@@ -51,12 +63,10 @@ in
 
         unitConfig.ConditionUser = "!root";
 
-        environment = {
-          # Default options
-          # https://gitlab.freedesktop.org/monado/monado/-/blob/4548e1738591d0904f8db4df8ede652ece889a76/src/xrt/targets/service/monado.in.service#L12
-          XRT_COMPOSITOR_LOG = mkDefault "debug";
-          XRT_PRINT_OPTIONS = mkDefault "on";
-          IPC_EXIT_ON_DISCONNECT = mkDefault "off";
+        environment = cfg.environment // {
+          XRT_COMPOSITOR_LOG = if builtins.hasAttr "XRT_COMPOSITOR_LOG" cfg.environment then cfg.environment.XRT_COMPOSITOR_LOG else "debug";
+          XRT_PRINT_OPTIONS = if builtins.hasAttr "XRT_PRINT_OPTIONS" cfg.environment then cfg.environment.XRT_PRINT_OPTIONS else "on";
+          IPC_EXIT_ON_DISCONNECT = if builtins.hasAttr "IPC_EXIT_ON_DISCONNECT" cfg.environment then cfg.environment.IPC_EXIT_ON_DISCONNECT else "off";
         };
```

Comment:
I think it would actually be better to set these values as defaults for `services.monado.environment`, so the module system can just do its job

Reply:
They are already set as the default there no? These if statements just set those values if they aren't independently defined

Reply:
If an attrs option has a default value, and you define a value, the default is going to be ignored. If you want to have persistent default values, you need to define them in the config block.

And even if you don't want to do that (many other modules do this), you can definitely simplify this expression by flipping the two operands of the update (`//`) operator.

Reply:
I have these defaults in the option block but that's more for documentation. I do want the persistent defaults so that's why I had the if statements in the config block.

Flipping the update operator definitely looks a lot better though:
```nix
environment = {
  XRT_COMPOSITOR_LOG = "debug";
  XRT_PRINT_OPTIONS = "on";
  IPC_EXIT_ON_DISCONNECT = "off";
} // cfg.environment;
```

Reply:
I meant, instead of settings these values in an expression, I would propose to define them as option values like this:

```nix
config = mkIf cfg.enable {
  services.monado.environment = {
    XRT_COMPOSITOR_LOG = lib.mkDefault "debug";
    XRT_PRINT_OPTIONS = lib.mkDefault "on";
    IPC_EXIT_ON_DISCONNECT = lib.mkDefault "off";
  };

  // ...
  systemd.user.services.monado.environment = cfg.environment;
}
```

Perhaps we could just alias `services.monado.environment` to `systemd.user.services.monado.environment`

Reply:
Example diff:

```diff
diff --git a/nixos/modules/services/hardware/monado.nix b/nixos/modules/services/hardware/monado.nix
index 0fb368814d2b..6384df84294e 100644
--- a/nixos/modules/services/hardware/monado.nix
+++ b/nixos/modules/services/hardware/monado.nix
@@ -4,10 +4,9 @@
 , ...
 }:
 let
-  inherit (lib) mkDefault mkEnableOption mkIf mkOption mkPackageOption mdDoc types;
+  inherit (lib) mkAliasOptionModule mkDefault mkEnableOption mkIf mkOption mkPackageOption types;
 
   cfg = config.services.monado;
-
 in
 {
   options.services.monado = {
@@ -29,21 +28,19 @@ in
 
     highPriority = mkEnableOption "high priority capability for monado-service"
       // mkOption { default = true; };
-
-    environment = mkOption {
-      type = with types; attrsOf (nullOr (oneOf [ str path package ]));
-      description = mdDoc "Environment variables passed to Monado.";
-      # Default options
-      # https://gitlab.freedesktop.org/monado/monado/-/blob/4548e1738591d0904f8db4df8ede652ece889a76/src/xrt/targets/service/monado.in.service#L12
-      default = {
-        XRT_COMPOSITOR_LOG = "debug";
-        XRT_PRINT_OPTIONS = "on";
-        IPC_EXIT_ON_DISCONNECT = "off";
-      };
-    };
   };
 
+  imports = [
+    (mkAliasOptionModule ["services" "monado" "environment"] ["systemd" "user" "services" "monado" "environment"])
+  ];
+
   config = mkIf cfg.enable {
+    services.monado.environment = {
+      XRT_COMPOSITOR_LOG = mkDefault "debug";
+      XRT_PRINT_OPTIONS = mkDefault "on";
+      IPC_EXIT_ON_DISCONNECT = mkDefault "off";
+    };
+
     security.wrappers."monado-service" = mkIf cfg.highPriority {
       setuid = false;
       owner = "root";
@@ -63,12 +60,6 @@ in
 
         unitConfig.ConditionUser = "!root";
 
-        environment = {
-          XRT_COMPOSITOR_LOG = "debug";
-          XRT_PRINT_OPTIONS = "on";
-          IPC_EXIT_ON_DISCONNECT = "off";
-        } // cfg.environment;
-
         serviceConfig = {
           ExecStart =
             if cfg.highPriority
```

Reply:
Yeah this looks like a better way to do it